# Segmenting and Clustering Neighborhoods in Toronto
<br>
<img src="https://cdn.pixabay.com/photo/2020/04/21/02/32/buildings-5070537_960_720.jpg" width="600" align="left" alt="Image of Toronto">

<a href="https://pixabay.com/photos/buildings-city-cityscape-skyscraper-5070537/">By lucasgeorgewendt from Pixabay</a>

### Import of necessary Libraries

In [1]:
!pip install geopy
!pip install folium
import pandas as pd
from geopy.geocoders import Nominatim
import folium
import requests
import numpy as np
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors
print('Libraries imported.')

Libraries imported.


## Task 1 - Dataframe Prep

In this section we scrape the necessary data of Toronto neighborhoods from Wikipedia and prepare it for our subsequent analysis.

### Scrape Tables from Webpage

Scraping of tables from the wikipedia link.

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# Load tables from URL
dfs = pd.read_html(url)

### Prepare Dataframe

Loading of the first table and removing the rows which include 'Not assigned'. Then reset of the index and display of the first 5 rows to check, whether or not the dataframe was imported correctly.

In [3]:
# Get first table                                                                                                           
df1 = dfs[0]

# Drop 'Not assigned' boroughs and reset index
df1.drop(df1.loc[df1['Borough']=='Not assigned'].index, inplace=True)

# Reset index
df1 = df1.reset_index(drop=True)

df1.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Show rows and columns of dataframe

In [4]:
df1.shape

(103, 3)

The dataframe has 103 rows and 3 columns.

## Task 2 -  Add Coordinates

In this section we add the longitude and latitude to our dataframe to be able to build cluster in task 3.

In [5]:
pCodes='http://cocl.us/Geospatial_data'

In [6]:
df2 = pd.read_csv(pCodes)
df2.head(5)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [7]:
df= df1.merge(df2, on=['Postal Code'])
df.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## Task 3 - Cluster Neighborhoods

Map creation of Toronto with all neighborhoods that are in Toronto main area. Then we cluster the neighborhoods based on defined criteria and display them on the map.

In [8]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tr_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Exclusion of all neighborhoods that do not contain Toronto in their borough name.

In [9]:
TO_df = df[df['Borough'].str.contains("Toronto")].reset_index(drop=True)
TO_df.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

### Map before clustering

Display of the neighborhoods before they are being clustered.

In [10]:
# define the world map centered around Toronto with an appropriate zoom level
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(TO_df['Latitude'], TO_df['Longitude'], TO_df['Borough'], TO_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  
toronto_map

### Venue import and merge with dataframe

Venue import for selected neighborhoods from Foursquare via API and merge with Toronto dataframe.

In [11]:
CLIENT_ID = 'SMB4TAZKRSNIXELA1GMJJ0PDHES3Y15Q1JTUUP2CDJURO1CM'
CLIENT_SECRET = 'TCCQVKRLMJ1BWJDXTI0VHDLDS2WRRA2ELGS35KGPH1IN5TMD'
VERSION = '20180605'
LIMIT = 50

Function for requesting venues from API.

In [12]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Return of function for neighborhoods in Toronto main area.

In [13]:
toronto_venues = getNearbyVenues(names=TO_df['Neighborhood'],
                                   latitudes=TO_df['Latitude'],
                                   longitudes=TO_df['Longitude']
                                  )

Check if function returned correct result.

In [14]:
print(toronto_venues.shape)
toronto_venues.head()

(1200, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

### One hot encoding and selection of 10 most common venues for k-means preparation

To prepare our data for being processed by the k-means algorithm the data will be prepared using one hot encoding and then the top 10 venues will be selected.

In [15]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

toronto_onehot.head()

Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   ...  Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0  ...                 0                 0      0              0   
1  ...                 0                 0      0              0   
2  ...                 0                 0      0              0   
3  ...                 0                 0      0              0   
4  ...                 0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Yoga Studio  
0         0            0            0  
1         0            0            0  
2         0            0            0  
3         0            0            0  
4         0            0            0  

[5 rows x 222 columns]

Calculate mean for one hot encoded venues.

In [16]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Airport  \
0                                        Berczy Park   0.0000   
1       Brockton, Parkdale Village, Exhibition Place   0.0000   
2  Business reply mail Processing Centre, South C...   0.0000   
3  CN Tower, King and Spadina, Railway Lands, Har...   0.0625   
4                                 Central Bay Street   0.0000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0              0.0000        0.0000           0.000            0.000   
1              0.0000        0.0000           0.000            0.000   
2              0.0000        0.0000           0.000            0.000   
3              0.0625        0.0625           0.125            0.125   
4              0.0000        0.0000           0.000            0.000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0             0.000                  0.0           0.0       0.0  ...   
1             0.000                  0.0           0.0       0.0  ...   
2             0.000                  0.0           0.0       0.0  ...   
3             0.125                  0.0           0.0       0.0  ...   
4             0.000                  0.0           0.0       0.0  ...   

   Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0               0.0               0.0    0.0            0.0   
1               0.0               0.0    0.0            0.0   
2               0.0               0.0    0.0            0.0   
3               0.0               0.0    0.0            0.0   
4               0.0               0.0    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                           0.02               0.0                    0.0   
1                           0.00               0.0                    0.0   
2                           0.00               0.0                    0.0   
3                           0.00               0.0                    0.0   
4                           0.02               0.0                    0.0   

   Wine Bar  Wings Joint  Yoga Studio  
0       0.0          0.0     0.000000  
1       0.0          0.0     0.000000  
2       0.0          0.0     0.055556  
3       0.0          0.0     0.000000  
4       0.0          0.0     0.020000  

[5 rows x 222 columns]

Function for only retrieving the 10 most common venues per neighborhood.

In [18]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [19]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant             Restaurant   
1           Coffee Shop        Breakfast Spot  Performing Arts Venue   
2           Yoga Studio         Auto Workshop             Skate Park   
3       Airport Service      Airport Terminal                Airport   
4        Sandwich Place    Italian Restaurant           Burger Joint   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery                  Café           Cheese Shop   
1                Bakery     Convenience Store             Pet Store   
2         Burrito Place        Farmers Market                   Spa   
3         Boat or Ferry   Rental Car Location      Sculpture Garden   
4        Ice Cream Shop   Japanese Restaurant                  Café   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar              Creperie         Breakfast Spot  
1          Climbing Gym            Restaurant          Burrito Place  
2  Fast Food Restaurant            Restaurant       Recording Studio  
3                 Plane           Coffee Shop        Harbor / Marina  
4       Bubble Tea Shop           Yoga Studio                  Hotel

### K-means clustering of neighborhoods

Clustering of neighborhoods into 9 clusters and subsequent illustration of clustered neighborhoods on map.

In [22]:
# set number of clusters
kclusters = 9

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering) 

In [24]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = TO_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               8           Coffee Shop   
1  43.662301 -79.389494               8           Coffee Shop   
2  43.657162 -79.378937               8                  Café   
3  43.651494 -79.375418               8                  Café   
4  43.676357 -79.293031               3                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1           Yoga Studio                  Bank         Smoothie Shop   
2           Coffee Shop             Bookstore            Restaurant   
3             Gastropub           Coffee Shop   American Restaurant   
4     Health Food Store                   Pub           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0        Breakfast Spot                  Café                      Theater   
1              Beer Bar        Sandwich Place                   Restaurant   
2      Ramen Restaurant              Tea Room               Clothing Store   
3        Cosmetics Shop          Cocktail Bar           Seafood Restaurant   
4         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0    Mexican Restaurant                 Shoe Store             Restaurant  
1         Burrito Place                       Café     College Auditorium  
2               Theater  Middle Eastern Restaurant         Cosmetics Shop  
3              Creperie          German Restaurant           Camera Store  
4   Dumpling Restaurant                 Donut Shop       Doner Restaurant

Plotting of formed clusters on Toronto map.

In [25]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters